### ISSUE TITLE
SparkWriter stages previously uploaded data

### ISSUE THREAD

**Issue by evgenystan** [snowflakedb/spark-snowflake#218]
I have a Spark Cluster that streams events from Kafka and saves them into a Snowflake table.
I process Kafka messages in batches every 30 seconds.
The batch processor is written in Python, however, it uses spark writer to save the updates to the table.
I am using this version of the Spark connector
`net.snowflake:spark-snowflake_2.11:2.7.0-spark_2.4,net.snowflake:snowflake-jdbc:3.12.3`
Code:
```
SNOWFLAKE_SOURCE_NAME = "net.snowflake.spark.snowflake"
sfOptions = {
    "sfURL": f"{SNOWFLAKE_DATA_CREDENTIAL['host']}.snowflakecomputing.com",
    "sfUser": SNOWFLAKE_DATA_CREDENTIAL['user'],
    "sfPassword": SNOWFLAKE_DATA_CREDENTIAL['password'],
    "sfDatabase": SNOWFLAKE_DATA_CREDENTIAL['database'],
    "sfSchema": SNOWFLAKE_DATA_CREDENTIAL['schema'],
    "sfWarehouse": SNOWFLAKE_DATA_CREDENTIAL['warehouse'],
    "dbtable": SNOWFLAKE_DATA_CREDENTIAL['table'],
    "sfRole": SNOWFLAKE_DATA_CREDENTIAL['role']
}
sf_df.write.format(SNOWFLAKE_SOURCE_NAME).mode('append')\
    .options(**sfOptions) \
    .save()
```
As I am watching the history of the queries in Snowflake, I see the data size constantly growing, and the row count increasing for each staged "COPY INTO" event.
```
command | size | driver | row count 
copy into TABLE FROM @spark_connector_load_stage_GMX92NImfX/...| 8.0MB | JDBC 3.12.3 | 117.4K
copy into TABLE FROM @spark_connector_load_stage_y7ZtmK7L2B... |8.0MB | JDBC 3.12.3 | 117.5K
```
I am confused as to why this is happening. I know for a fact that the data I am writing is tiny in the dev environment, as you can see only about hundred rows was added to the staged table.

Yes, only the updates are written to the Snowflake table, but this is probably going to blow up, when I move my code to production, and the data will quickly accumulate into gigabytes.

Note: setting the option `USE_CACHED_RESULT` to `False` does nothing.

**Comment by evgenystan** [comment]
I have a felling that this is the perpetrator:
https://github.com/snowflakedb/spark-snowflake/blob/master/src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala#L1334

I have a hunch that the internal S3 stage (we have Snowflake deployed on AWS) has all files dumped in one location, and at each `COPY INTO` rereads all previous files.

I may be wrong, but other than that, I couldn't fine any other place in the code, where such data duplication may occur and go unnoticed.

Yes, temporary stages are supposed to have files cleaned up upon session close, but I suspect that cleanup is not happening, and there is no place in the code that cleans up internal S3 storage.

**Comment by evgenystan** [comment]
Ok here we go, I think I've found it.

Spark writer creates a temporary stage.
Then it executes `GET` for some mysterious reason.
Then it executes `PUT`:
```
PUT file:///tmp/dummy_location_spark_connector_tmp/ @spark_connector_load_stage_HpS705tCYB
```
Then it executes the copy statement from a subdirectory...
```
copy into SECURITY_EVENT FROM @spark_connector_load_stage_HpS705tCYB/gJpP9wM7UW 
FILE_FORMAT = (
    TYPE=CSV
    FIELD_DELIMITER='|'
    NULL_IF=()
    FIELD_OPTIONALLY_ENCLOSED_BY='"'
    TIMESTAMP_FORMAT='TZHTZM YYYY-MM-DD HH24:MI:SS.FF3'
    DATE_FORMAT='TZHTZM YYYY-MM-DD HH24:MI:SS.FF3'
  )
```
The thing is, this subdirectory contains all files that were created by the `UploadRDD` function here 
https://github.com/snowflakedb/spark-snowflake/blob/caeafde3c8db65c218b115f839b04751b78387bf/src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala#L681
This function creates the 10 character random alphanumeric name for the directory and these files will be later copied to the temporary directory `file:///tmp/dummy_location_spark_connector_tmp/ ` where all previous subdirectories reside
The `PUT` command is formed here 
https://github.com/snowflakedb/spark-snowflake/blob/caeafde3c8db65c218b115f839b04751b78387bf/src/main/scala/net/snowflake/spark/snowflake/io/SFInternalStage.scala#L154
And that command uploads all files residing in this directory and all subdirectories... Voila!!!

I will have to implement a temporary fix that will clean the temporary directory before executing the `writer.save()` but will appreciate you fixing it in code.

**Comment by rkesh-singh** [comment]
@evgenystan can you help me to understand where exactly file:///tmp/dummy_location_spark_connector_tmp/ this directory is created, is it in spark cluster or Snowflake cluster?

**Comment by evgenystan** [comment]
@rkesh-singh 
This location does not exist anywhere. If `tempDir` option is not provided, SparkConnector uploads the files directly to an internal Snowflake S3 stage from a byte stream created in memory. Yes, the connector executes these `GET` and `PUT` commands, but it looks like these commands do nothing, because the files are never on EMR cluster.
And where the files reside at internal stage?.. I don't know, I did not try to find out.

So I tried forcing external staging by specifying `tempDir` option. This allowed me to see files. Which are less than a kilobyte in size for each batch. Yet the history logs still showed megabytes of data and thousands of rows being processed.

I also noticed that these history statistics do go down from time to time, so it seems that Snowflake does clean up internal stage periodically.

I deployed my code in production, and this is not happening in production (as far as I can tell now). Probably due to the size of the files or more diligent cleanup.
